### PR TITLE
Signals changes from sync review and WG3

### PR DESCRIPTION
**Pull request by sadiqj**
This PR implements the three changes coming out of @xavierleroy 's review of the signal code and the comments from the runtime working group:

1. We now block all signals before spawning a domain and then unblock afterwards immediately and when safe to do so for the parent and child respectively. This is https://github.com/ocaml-multicore/ocaml-multicore/commit/3b1db34d0f84ac44d5382bcb638f1e482cc4c7f5
2. We have removed the `total_signals_pending`. This makes some things a little more expensive but the intention is we could potentially optimise these with bitvector tricks or suchlike at a later stage. This is https://github.com/ocaml-multicore/ocaml-multicore/commit/7b389fa4b980732825d4b0fc67b9a18b4190934e
3. We now coalesce signals by signal number. This is https://github.com/ocaml-multicore/ocaml-multicore/commit/5dc58d93ac82021f6771d9cfb9725f8fe46302f7
4. Suspicious indentation in `caml_process_pending_signals_exn` is fixed. This is https://github.com/ocaml-multicore/ocaml-multicore/commit/8de958914bea7188b9a56a158a2762f81f518305 (https://github.com/ocaml-multicore/ocaml-multicore/issues/742)

### CODE REVIEW

File: runtime/domain.c
Comment (xavierleroy) [inline runtime/domain.c]: It's sad to see so many  `#ifdef POSIX_SIGNALS`.  If you have POSIX threads, you have POSIX signals.  Since you're calling `pthread_create` directly, you have POSIX threads...

The only platform that does not have POSIX_SIGNALS is Win32.  On the one hand it's good to make provisions for a future Win32 port.  On the other hand a lot of the current code assumes POSIX threads and doesn't work with Win32.

File: runtime/signals.c
Comment (xavierleroy) [inline runtime/signals.c:94]: You might be able to simplify the whole `again: ... goto again` by using a single atomic_exchange, to read the old value and `caml_pending_signals[i]` and set it to 0 atomically.  On the other hand, the current code might be more efficient if `caml_pending_signals[i]` is zero most of the time, which it should be in practice.  Your choice.

**Review by xavierleroy — COMMENTED** [review]
The switch to Boolean-valued signals and the removal of `total_signals_pending` look good to me.  I need to re-read the signal-blocking material in context, because I got lost just reading the diff.

File: runtime/domain.c
Comment (sadiqj) [reply runtime/domain.c]: Yes, this is a good point. So much of `domain.c` will have to change for Win32 that it's better to just assume POSIX_SIGNALS. I have removed the ifdefs.

File: runtime/signals.c
Comment (sadiqj) [reply runtime/signals.c:94]: Exchange does simplify the code quite nicely but having a ponder I think CAS is better here. I worry about multiple sockets playing cache line ping-pong as a few hundred threads come out of a stop-the-world pause and all race to deliver a waiting signal.

**Comment by abbysmal** [comment]
The changes looks good to me, there's however one omission I believe: `caml_thread_new` should be masking during thread creation as well. (in `st_stubs.c`).

Otherwise everything looks good to me.

**Comment by sadiqj** [comment]
> The changes looks good to me, there's however one omission I believe: `caml_thread_new` should be masking during thread creation as well. (in `st_stubs.c`).

Good catch. Will fix that.

**Comment by sadiqj** [comment]
@Engil this should block for `caml_thread_new` as well now. When we add win32 support we should consider moving the logic to `st_create_thread`.

**Comment by abbysmal** [comment]
LGTM now, thanks.
@xavierleroy you mentioned wanting to re-read the signal-blocking material in context, should we merge this PR as in (and have a wider look at it during merge) or wait for further review from your part?

**Review by abbysmal — APPROVED** [review]

File: runtime/signals.c
Comment (xavierleroy) [inline runtime/signals.c]: As far as I can see, it's enough to atomically store 1 in `caml_pending_signals[signal_number]`.  Plus, this avoids the highly theoretical risk of overflow and wrapping around to 0...

**Review by xavierleroy — APPROVED** [review]
I downloaded the branch and read the changes again but with more context this time.  It all looks good to me.  Just a suggestion for one more simplification below.

File: runtime/signals.c
Comment (sadiqj) [reply runtime/signals.c]: Yes, this makes sense. Changing.

**Comment by abbysmal** [comment]
Everything seems fine now, thank you everyone for your reviews.
Merging.

**Pull request merged**
